Regression: Workspaces aren't persisted due to missing JSON converter

```
00:00:03.815 [ERROR] Exception while loading window state|System.InvalidOperationException: Use LoadoutId.From instead.
   at NexusMods.Abstractions.Loadouts.Ids.LoadoutId..ctor() in /ssd-pool/projects/NexusMods/app/main/src/Abstractions/NexusMods.Abstractions.Loadouts/TransparentValueObjects/TransparentValueObjects.ValueObjectIncrementalSourceGenerator/NexusMods.Abstractions.Loadouts.Ids.LoadoutId.g.cs:line 68
   at .ctor()
   at System.Text.Json.Serialization.Converters.ObjectDefaultConverter`1.OnTryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, T& value)
   at System.Text.Json.Serialization.JsonConverter`1.TryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, T& value, Boolean& isPopulatedValue)
   at System.Text.Json.Serialization.JsonConverter`1.ReadCore(Utf8JsonReader& reader, JsonSerializerOptions options, ReadStack& state)
   at System.Text.Json.Serialization.Metadata.JsonTypeInfo`1.Deserialize(Utf8JsonReader& reader, ReadStack& state)
   at System.Text.Json.JsonSerializer.Read[TValue](Utf8JsonReader& reader, JsonTypeInfo`1 jsonTypeInfo)
   at System.Text.Json.JsonSerializer.Deserialize[TValue](Utf8JsonReader& reader, JsonSerializerOptions options)
   at lambda_method18(Closure, Utf8JsonReader&, Type, JsonSerializerOptions)
   at NexusMods.Abstractions.Serialization.Json.AExpressionConverterGenerator`1.Read(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options) in /ssd-pool/projects/NexusMods/app/main/src/Abstractions/NexusMods.Abstractions.Serialization/Json/AExpressionConverterGenerator.cs:line 37
   at System.Text.Json.Serialization.JsonConverter`1.TryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, T& value, Boolean& isPopulatedValue)
   at System.Text.Json.Serialization.JsonConverter`1.ReadCore(Utf8JsonReader& reader, JsonSerializerOptions options, ReadStack& state)
   at System.Text.Json.Serialization.Metadata.JsonTypeInfo`1.DeserializeAsObject(Utf8JsonReader& reader, ReadStack& state)
   at System.Text.Json.JsonSerializer.ReadAsObject(Utf8JsonReader& reader, JsonTypeInfo jsonTypeInfo)
   at NexusMods.Abstractions.Serialization.Json.AbstractClassConverterGenerator`1.Read(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options) in /ssd-pool/projects/NexusMods/app/main/src/Abstractions/NexusMods.Abstractions.Serialization/Json/AbstractConverterGenerator.cs:line 90
   at System.Text.Json.Serialization.JsonConverter`1.TryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, T& value, Boolean& isPopulatedValue)
   at System.Text.Json.Serialization.Metadata.JsonPropertyInfo`1.ReadJsonAndSetMember(Object obj, ReadStack& state, Utf8JsonReader& reader)
   at System.Text.Json.Serialization.Converters.ObjectDefaultConverter`1.OnTryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, T& value)
   at System.Text.Json.Serialization.JsonConverter`1.TryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, T& value, Boolean& isPopulatedValue)
   at System.Text.Json.Serialization.JsonCollectionConverter`2.OnTryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, TCollection& value)
   at System.Text.Json.Serialization.JsonConverter`1.TryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, T& value, Boolean& isPopulatedValue)
   at System.Text.Json.Serialization.Metadata.JsonPropertyInfo`1.ReadJsonAndSetMember(Object obj, ReadStack& state, Utf8JsonReader& reader)
   at System.Text.Json.Serialization.Converters.ObjectDefaultConverter`1.OnTryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, T& value)
   at System.Text.Json.Serialization.JsonConverter`1.TryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, T& value, Boolean& isPopulatedValue)
   at System.Text.Json.Serialization.JsonConverter`1.ReadCore(Utf8JsonReader& reader, JsonSerializerOptions options, ReadStack& state)
   at System.Text.Json.JsonSerializer.ReadFromSpan[TValue](ReadOnlySpan`1 utf8Json, JsonTypeInfo`1 jsonTypeInfo, Nullable`1 actualByteCount)
   at System.Text.Json.JsonSerializer.ReadFromSpan[TValue](ReadOnlySpan`1 json, JsonTypeInfo`1 jsonTypeInfo)
   at NexusMods.App.UI.Windows.WindowDataAttributes.Model.get_Data() in /ssd-pool/projects/NexusMods/app/main/src/NexusMods.App.UI/Windows/WindowDataAttributes.cs:line 31
   at NexusMods.App.UI.Windows.WindowManager.RestoreWindowState(IWorkspaceWindow window) in /ssd-pool/projects/NexusMods/app/main/src/NexusMods.App.UI/Windows/WindowManager.cs:line 122

```

--- a/src/Abstractions/NexusMods.Abstractions.Loadouts/Ids/FileId.cs
+++ b/src/Abstractions/NexusMods.Abstractions.Loadouts/Ids/FileId.cs
@@ -9,7 +9,7 @@ namespace NexusMods.Abstractions.Loadouts.Ids;
 /// then one winning file is written to disk when the loadout is applied.
 /// </summary>
 [ValueObject<EntityId>]
-public readonly partial struct FileId : ITypedId<File.Model>
+public readonly partial struct FileId : ITypedId<File.Model>, IAugmentWith<JsonAugment>
 {
     EntityId ITypedEntityId.Value => Value;
 }

--- a/src/Abstractions/NexusMods.Abstractions.Loadouts/Ids/LoadoutId.cs
+++ b/src/Abstractions/NexusMods.Abstractions.Loadouts/Ids/LoadoutId.cs
@@ -2,7 +2,6 @@ using JetBrains.Annotations;
 using NexusMods.Abstractions.MnemonicDB.Attributes.Extensions;
 using NexusMods.MnemonicDB.Abstractions;
 using TransparentValueObjects;
-using static NexusMods.Abstractions.MnemonicDB.Attributes.Extensions.EntityExtensions;
 
 namespace NexusMods.Abstractions.Loadouts.Ids;
 
@@ -14,7 +13,7 @@ namespace NexusMods.Abstractions.Loadouts.Ids;
 /// </summary>
 [ValueObject<EntityId>]
 [PublicAPI]
-public readonly partial struct LoadoutId : ITypedId<Loadout.Model>
+public readonly partial struct LoadoutId : ITypedId<Loadout.Model>, IAugmentWith<JsonAugment>
 {
     
     /// <inheritdoc />

--- a/src/Abstractions/NexusMods.Abstractions.Loadouts/Ids/ModId.cs
+++ b/src/Abstractions/NexusMods.Abstractions.Loadouts/Ids/ModId.cs
@@ -9,7 +9,7 @@ namespace NexusMods.Abstractions.Loadouts.Ids;
 /// they are from the same installation archive.
 /// </summary>
 [ValueObject<EntityId>]
-public readonly partial struct ModId : ITypedId<Mod.Model>
+public readonly partial struct ModId : ITypedId<Mod.Model>, IAugmentWith<JsonAugment>
 {
     EntityId ITypedEntityId.Value => Value;
 }

--- a/src/Abstractions/NexusMods.Abstractions.Serialization/Json/EntityIdConverter.cs
+++ b/src/Abstractions/NexusMods.Abstractions.Serialization/Json/EntityIdConverter.cs
@@ -1,0 +1,24 @@
+using System.Text.Json;
+using System.Text.Json.Serialization;
+using JetBrains.Annotations;
+using NexusMods.MnemonicDB.Abstractions;
+
+namespace NexusMods.Abstractions.Serialization.Json;
+
+/// <inheritdoc/>
+[UsedImplicitly]
+public class EntityIdConverter : JsonConverter<EntityId>
+{
+    /// <inheritdoc/>
+    public override EntityId Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+    {
+        var value = reader.GetUInt64();
+        return EntityId.From(value);
+    }
+
+    /// <inheritdoc/>
+    public override void Write(Utf8JsonWriter writer, EntityId value, JsonSerializerOptions options)
+    {
+        writer.WriteNumberValue(value.Value);
+    }
+}

--- a/src/Abstractions/NexusMods.Abstractions.Serialization/NexusMods.Abstractions.Serialization.csproj
+++ b/src/Abstractions/NexusMods.Abstractions.Serialization/NexusMods.Abstractions.Serialization.csproj
@@ -9,6 +9,7 @@
     </ItemGroup>
 
     <ItemGroup>
+        <ProjectReference Include="..\NexusMods.Abstractions.MnemonicDB.Attributes\NexusMods.Abstractions.MnemonicDB.Attributes.csproj" />
         <ProjectReference Include="..\NexusMods.Abstractions.Settings\NexusMods.Abstractions.Settings.csproj" />
     </ItemGroup>
 

--- a/src/Abstractions/NexusMods.Abstractions.Serialization/Services.cs
+++ b/src/Abstractions/NexusMods.Abstractions.Serialization/Services.cs
@@ -19,7 +19,8 @@ public static class Services
         services.AddSettingsStorageBackend<JsonStorageBackend>();
 
         services.AddSingleton<JsonConverter, AbstractClassConverterFactory<IMetadata>>();
-        
+        services.AddSingleton<JsonConverter, EntityIdConverter>();
+
         services.AddSingleton(s =>
         {
             var opts = new JsonSerializerOptions();

--- a/src/NexusMods.App.UI/Windows/WindowDataAttributes.cs
+++ b/src/NexusMods.App.UI/Windows/WindowDataAttributes.cs
@@ -19,23 +19,18 @@ public class WindowDataAttributes
 
     public class Model(ITransaction tx) : Entity(tx)
     {
-        /// <summary>
-        /// The data of the window
-        /// </summary>
-        public WindowData Data
+        public WindowData GetData()
         {
-            get
-            {
-                var options = Db.Connection.ServiceProvider.GetRequiredService<JsonSerializerOptions>();
-                var jsonString = WindowDataAttributes.Data.Get(this);
-                return JsonSerializer.Deserialize<WindowData>(jsonString, options)!;
-            }
-            set
-            {
-                var options = Db.Connection.ServiceProvider.GetRequiredService<JsonSerializerOptions>();
-                var jsonString = JsonSerializer.Serialize(value, options);
-                WindowDataAttributes.Data.Add(this, jsonString);
-            }
+            var options = Db.Connection.ServiceProvider.GetRequiredService<JsonSerializerOptions>();
+            var jsonString = Data.Get(this);
+            return JsonSerializer.Deserialize<WindowData>(jsonString, options)!;
+        }
+
+        public void SetData(WindowData value)
+        {
+            var options = Db.Connection.ServiceProvider.GetRequiredService<JsonSerializerOptions>();
+            var jsonString = JsonSerializer.Serialize(value, options);
+            Data.Add(this, jsonString);
         }
     }
 }

--- a/src/NexusMods.App.UI/Windows/WindowManager.cs
+++ b/src/NexusMods.App.UI/Windows/WindowManager.cs
@@ -93,16 +93,17 @@ internal sealed class WindowManager : ReactiveObject, IWindowManager
             using var tx = _conn.BeginTransaction();
             if (!_repository.TryFindFirst(out var found))
             {
-                _ = new WindowDataAttributes.Model(tx)
+                var model = new WindowDataAttributes.Model(tx)
                 {
                     Db = _conn.Db,
-                    Data = data,
                 };
+
+                model.SetData(data);
             }
             else
             {
                 found.Tx = tx;
-                found.Data = data;
+                found.SetData(data);
             }
             tx.Commit();
         }
@@ -119,7 +120,7 @@ internal sealed class WindowManager : ReactiveObject, IWindowManager
             if (!_repository.TryFindFirst(out var found))
                 return false;
 
-            window.WorkspaceController.FromData(found.Data);
+            window.WorkspaceController.FromData(found.GetData());
             return true;
         }
         catch (Exception e)

